### PR TITLE
cmd/snap: use updated "current" revision after snap refresh run inhibition

### PR DESCRIPTION
**Pull request by ZeyadYasser**
This is an alternative to https://github.com/snapcore/snapd/pull/13343 that addresses race conditions discussed.

This approach is different from https://github.com/snapcore/snapd/pull/13410 where here the lock is kept as long as possible (i.e. until just before running snap-confine, check `beforeExec`).

~~**NOTE: Please ignore all test related code as it will need serious refactoring which should be postponed until after we decide on an approach.**~~

### CODE REVIEW

File: cmd/snap/cmd_run.go
Comment (pedronis) [inline cmd/snap/cmd_run.go:521]: any reason not to wait already here in this approach? which would also mean maybe we don't need the changes to remember the snap revision? or do we have other places where we need that?

File: cmd/snap/inhibit.go
Comment (pedronis) [inline cmd/snap/inhibit.go:150]: this is dealing too directly with the lock, so probably it should live as a helper in snaplock/runihnibit taking callbacks to deal with the notification flow etc

File: cmd/snap/cmd_run.go
Comment (pedronis) [inline cmd/snap/cmd_run.go:524]: notice that this code as is as the same problem as the original PR where the current symlink might not available at this point, another reason to wait before this point

File: cmd/snap/inhibit.go
Comment (pedronis) [inline cmd/snap/inhibit.go:145]: it probably needs to be clearer that this returns holding the lock

**Review by pedronis — COMMENTED** [review]
I looked at this, some of it is reasonable but I have questions because exactly as it is it doesn't quite solve the problem we set out to solve originally

File: cmd/snap/cmd_run.go
Comment (pedronis) [reply cmd/snap/cmd_run.go:521]: no, maybe we want that logic, but the main point about needing to hold the lock to get the info with snap.R(0) and needing to wait/grab the lock earlier stands

Comment (ZeyadYasser) [reply cmd/snap/cmd_run.go:521]: This is done later in runSnapConfine here: https://github.com/snapcore/snapd/pull/13411/files#diff-8362a8b80bc8ecb25f0a01a694035de268e569287891db10dfd0426c1a411f25R1062-R1076

File: cmd/snap/cmd_run.go
Comment (pedronis) [inline cmd/snap/cmd_run.go:1069]: I don't think we can use defer to finish the flow, as at the end of this chain of functions we do exec which will not return, no?

**Review by pedronis — CHANGES_REQUESTED** [review]
another remark about the current code. A sketch of what I have in mind is something like this (not tested, needs some more details):

```
// lives in snaplock/runinhibit, returns a held read lock on success
func waitWhileInhibited(snapName string, notInhibited func() error, inhibited func(*runinhibit.InhibitInfo) (cont bool, err error)) (flock *osutil.FileLock, err error) {
	...
}

func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
	if x.DebugLog {
		os.Setenv("SNAPD_DEBUG", "1")
		logger.Debugf("enabled debug logging of early snap startup")
	}
	snapName, appName := snap.SplitSnapApp(snapApp)

	var snapInfo *snap.Info
	var app *snap.AppInfo
	first := true

	notInhibited := func() (err error) {
		snapInfo, app, err = getInfoAndApp(snapName, appName, snap.R(0))
		return err
	}
	inhibited := func(inhibitInfo *runinhibit.InhibitInfo) (cont bool, err error) {
		snapInfo, app, err = getInfoAndApp(snapName, appName, inhibitInfo.Previous)
		if err != nil {
			return false, err
		}
		if app.IsService() {
			// cont == true means we should not wait
			return true, nil
		}
		if first {
			// start flow ...
			first = false
		}
		return false, nil
	}
	// XXX wrap it in maybe... to check for the experimental flag
	flock, err := waitWhileInhibited(snapName, notInhibited, inhibited)
	if err != nil {
		return err
	}
	// ... setup to stop flow and close flock as needed

	return x.runSnapConfine(snapInfo, app.SecurityTag(), snapApp, "", args)
}
```

**Comment by ZeyadYasser** [comment]
@pedronis @olivercalder going back to this discussion https://github.com/snapcore/snapd/pull/13343#discussion_r1408181027

It seems like we cannot create the lock file in `snap run` as non-root. I have an approach in mind but it might not be the cleanest, your suggestions would be much appreciated.

We setup an empty inhibit lock file on first install for each snap (e.g. inside `link-snap`), and for existing snap we check for missing inhibit lock files and create them at startup?

File: cmd/snaplock/runinhibit/inhibit.go
Comment (olivercalder) [inline cmd/snaplock/runinhibit/inhibit.go:307]: Something like:
```go
// The notInhibited callback is called with the hint file lock is acquired if the snap is not inhibited.
// The inhibited callback is called with the hint file lock held every time the lock is acquired and the snap is still inhibited.
// If inhibited returns cont==false, it waits for the given interval and retries to acquire the lock.
// Returns with the hint file lock held.
```

But that may be too much detail, not sure. I think it's useful to point out that either the `inhibited` callback is called or the `notInhibited` callback, but not both, unless the `inhibited` callback signals to retry.

File: cmd/snaplock/runinhibit/inhibit.go
Comment (olivercalder) [inline cmd/snaplock/runinhibit/inhibit.go:349]: Technically, don't need this else here, as the if returns.

File: cmd/snaplock/runinhibit/inhibit.go
Comment (olivercalder) [inline cmd/snaplock/runinhibit/inhibit.go:326]: I don't think `retErr` is set elsewhere, but I think all the error cases in the for loop need to set it. Perhaps that would be an argument for making the returned `retErr` simply called `err`. Or could explicitly rename all the other ones so it's clear that they set the return error.

Perhaps you're right though, it'll assign whatever err is returned to `retErr`, and it's not necessary to do so explicitly. This is one of those scary Go things to me, but up to you.

File: cmd/snap/cmd_run.go
Comment (olivercalder) [reply cmd/snap/cmd_run.go:521]: (later me here) the `inhibited` callback could be called multiple times. Why don't you want to call the notification flow more than once, if it gets re-inhibited again? I assume we want to let the user know somehow that the snap is inhibited again, so they're not sitting there staring at a completed progress bar.

**Review by olivercalder — COMMENTED** [review]
Thanks, this looks really great! I have a few questions, mostly for my own understanding though, or nitpicks -- sorry if it's a bit much!

I the only blockers to my approving are whether the notify flow should run more than once, and if there's no inhibit race between when the lock is finally released and when the command is executed. And perhaps whether `beforeExec` should be named something to indicate that it releases the hint file lock, if that's a thing that it always does.

**Comment by olivercalder** [comment]
> We setup an empty inhibit lock file on first install for each snap (e.g. inside `link-snap`), and for existing snap we check for missing inhibit lock files and create them at startup?

I like this approach. Unless we use the fact that the inhibit lock file doesn't exist to check whether a snap has ever been run, I think it's good to assume that an inhibit lock ought to exist, and to create one on install.

What if it is deleted directly, though -- what is the fallback behavior? And how are they otherwise created at the moment, if running the snap doesn't create them (and can't)?

File: cmd/snap/cmd_run.go
Comment (olivercalder) [reply cmd/snap/cmd_run.go:492]: In particular, it seems like it could be re-created on every invocation of `inhibited()`, and even then only if `!notified` (assuming you only want to call it once, which I'm not sure about).

File: cmd/snap/cmd_run.go
Comment (ZeyadYasser) [reply cmd/snap/cmd_run.go]: You are correct that `inhibited` callback can be called multiple times for the same inhibition, this is why it is only allowed to run once to avoid spamming the user. It is very unlikely that a snap gets refreshed twice in < 1 second, but if it does then it wouldn't hurt if the user kept waiting for the second refresh to complete, I think it might even make sense from a UX perspective to do that. I feel it might be a bit confusing if I get two consecutive flows for two different refreshes from a single snap run.

Thank you for bringing this up, as it will definitely be helpful when doing the RAA UX review. I think these edge cases can be better addressed in the `Inhibition Flows` once a proper UX review is conducted to avoid cluttering snap run logic further

I might be misunderstanding your point, please let me know if I didn't address your questions?

File: cmd/snap/cmd_run_test.go
Comment (ZeyadYasser) [reply cmd/snap/cmd_run_test.go:247]: I completely agree with you, it currently has a single purpose and it makes sense to name it accordingly. I was aiming for something generic because it is being used also in `snapRunHook` and the callback now either closes the flock or triggers a retry so the logic can grow.

File: cmd/snap/inhibit.go
Comment (ZeyadYasser) [reply cmd/snap/inhibit.go:143]: > Perhaps var runinhibitWaitWhileInhibited = runinhibit.WaitWhileInhibited would achieve both?

I really like this approach, Thanks!

File: cmd/snap/cmd_run_test.go
Comment (ZeyadYasser) [reply cmd/snap/cmd_run_test.go:682]: Good point, I also noticed this and thought it was odd, I was trying to follow same patterns in the file, but you are absolutely right. Since it was not introduced by this PR, maybe it is better to do a separate cleanup PR to refactor it?

File: cmd/snaplock/runinhibit/inhibit.go
Comment (ZeyadYasser) [reply cmd/snaplock/runinhibit/inhibit.go:371]: Great catch, Thank you! Do you think switching to sleep instead of a ticker would be more reasonable?

File: cmd/snaplock/runinhibit/inhibit.go
Comment (ZeyadYasser) [reply cmd/snaplock/runinhibit/inhibit.go:330]: I am using a separate name intentionally here for readability, `err` will work though but I wanted to avoid the confusion about which error is read due to scopes.

Named return values are assigned before defer calls: https://go.dev/blog/defer-panic-and-recover
> 3. Deferred functions may **read** and assign to the returning function’s named return values.

Comment (ZeyadYasser) [reply cmd/snaplock/runinhibit/inhibit.go:330]: `TestWaitWhileInhibitedCallbackError` exercises this cleanup defer call.

File: cmd/snap/cmd_run.go
Comment (ZeyadYasser) [reply cmd/snap/cmd_run.go:1355]: AFAICT, you are right, but I don't see how we can workaround this. @pedronis would love to know your opinion.

**Comment by ZeyadYasser** [comment]
> I like this approach. Unless we use the fact that the inhibit lock file doesn't exist to check whether a snap has ever been run, I think it's good to assume that an inhibit lock ought to exist, and to create one on install.
> 
> What if it is deleted directly, though -- what is the fallback behavior? And how are they otherwise created at the moment, if running the snap doesn't create them (and can't)?

@olivercalder After some discussion with @pedronis, a retry mechanism was adopted to handle absent hint files.

File: cmd/snap/cmd_run.go
Comment (ZeyadYasser) [inline cmd/snap/cmd_run.go:1352]: @pedronis should we be retrying after transient cgroups are created? would we be creating multiple cgroups for the same process, hence leaking cgroups?

**Review by pedronis — COMMENTED** [review]
it feels like the additions to snaplock/runinhibit coudl be in their own separate PR? this PR is again quite big

File: cmd/snap/cmd_run.go
Comment (pedronis) [reply cmd/snap/cmd_run.go:1354]: as I said we could explore removing the cgroup if we need to retry, related to the leaking we could also remember that we created it already and not doing it again

File: cmd/snap/cmd_run.go
Comment (olivercalder) [reply cmd/snap/cmd_run.go]: Makes sense, thank you! I guess it's a UX question whether it makes more sense for a single flow to run, then wait again while second silently happens, or for two flows to run in this (very unlikely) case. I think I favor the former, at least if the refreshes don't overlap, which I don't think they should. And if refreshes take >= 1 second, which I'm not sure is always the case. If a refresh could take a long time, due to slow network or something, I think there should be more than one flow.

File: cmd/snap/cmd_run.go
Comment (olivercalder) [inline cmd/snap/cmd_run.go:536]: I see, this is why the `flow` needs to be defined outside of the `inhibited` callback. Thanks!

File: cmd/snap/cmd_run_test.go
Comment (olivercalder) [reply cmd/snap/cmd_run_test.go:682]: It's probably not important enough to fix, I was just curious. Thanks though!

File: cmd/snaplock/runinhibit/inhibit.go
Comment (olivercalder) [reply cmd/snaplock/runinhibit/inhibit.go:371]: I think sleep would be a bit safer, yeah, just to guarantee there's actually time for it to possibly become uninhibited. It might make sense to scale this based on architecture (see `testutil.HostScaledTimeout`), but that's only if meaningful work has to be done with it unlocked. If the work to uninhibit the snap just involves waiting to acquire the lock, then take all the time it needs, then release, there should be no reason to scale the time slept here.

Hope this makes sense..... Basically I think sleep should probably be fine, and no need to scale sleep time based on architecture.

File: cmd/snaplock/runinhibit/inhibit.go
Comment (olivercalder) [reply cmd/snaplock/runinhibit/inhibit.go:330]: Very good point, thanks for explaining!

File: cmd/snap/cmd_run_test.go
Comment (olivercalder) [reply cmd/snap/cmd_run_test.go:247]: Yes, with the retry, I think the generic name is better.

File: cmd/snap/cmd_run.go
Comment (olivercalder) [reply cmd/snap/cmd_run.go:1357]: This is the big question for me still, before approving.

**Review by olivercalder — COMMENTED** [review]
To clarify my earlier review, this looks generally good, the lingering question for me is how we ensure that the snap doesn't become inhibited again right as we're about to run it.

File: cmd/snap/cmd_run.go
Comment (pedronis) [reply cmd/snap/cmd_run.go:1357]: yes, it's about the cgroup, snapd unless forced will not inhibit the snap and do a refresh anymore because it detects app processes running for via the established cgroup, also notice that the lock would go away anyway on the exec as we open our locks with O_CLOEXEC

**Comment by ZeyadYasser** [comment]
Marking as blocked until https://github.com/snapcore/snapd/pull/13435 gets merged.

**Comment by ZeyadYasser** [comment]
@pedronis @olivercalder I created a couple of smaller PRs in attempt to break this PR down further and keeping it focused and easier to review
- https://github.com/snapcore/snapd/pull/13467 for refactoring notification flows.
- https://github.com/snapcore/snapd/pull/13468 for transient cgroup tracking.

In the smaller PRs, I am linking back to this PR for context on their purpose.

**Comment by ZeyadYasser** [comment]
Closing in favor of https://github.com/snapcore/snapd/pull/13747.

A lot changed on master and a rebase would have been very messy, I thought it would be better to keep the discussions here intact with the changes history and created a new PR instead.

**Pull request closed without merging**
